plio/io/io_controlnetwork.py

```diff
@@ -211,6 +278,10 @@ def read(self):
         if 'aprioriline' in df.columns:
             df['aprioriline'] -= 0.5
             df['apriorisample'] -= 0.5
+
+        # Munge the MeasureLogData into Python objs
+        df['measureLog'] = df['measureLog'].apply(lambda x: [Log.from_protobuf(i) for i in x])
```

Comment:
I am confused by this line. I get that the apply is happening to every row of the 'measureLog' data series..but can you grab a class method directly off of a class definition? Log wasn't instantiated anywhere, but it is a defined class, although not the current class we are in. Additionally, where is x defined? How can the list comprehension grab from it?

Reply:
This is the best explanation I have found about [class methods](https://stackoverflow.com/questions/12179271/meaning-of-classmethod-and-staticmethod-for-beginner) in python. Basically, yes. The method is associated with the class and returns an instance of the class.

In the lambda the 'x' is entirely arbitrary. So the lambda here takes one argument (called whatever) and passes that into the function. The function then returns (without a `return`) the list comprehension (in the instance a list of MeasureLog Objects). We have an example of this on the cost function for subpixel as well except the cost takes two variables (randomly named x and y).

We could write the same thing as:

```
def create_logs_from_iterable(iterable):
    return [MeasureLog.from_protobuf(x) for x in iterable]

# I think the axis=1 is needed in this form...
df['measureLog'].apply(create_logs_from_iterable, axis=1)
```